### ISSUE TITLE
French BMS - import samples and occurrences

### ISSUE THREAD

**Issue by andrewvanbreda** [BiologicalRecordsCentre/ABLE#658]
First task is to work with Mathieu and Benoît to prepare data so it is in suitable format for import.

Locations have already been imported here https://github.com/BiologicalRecordsCentre/ABLE/issues/514

**Comment by andrewvanbreda** [comment]
Important note from Jim

"If historic data is imported, it is important not to simply join on the original section_id (stored in the external_key) as it was non-unique. A combination of transect_id (also stored in the external_key) and the section_id must be used."

This sounds like it means that if we are going to use the Indicia data importer it will likely be necessary to pre-collect the Indicia location IDs for the sections to attach the samples to, as I don't think the Indicia importer can be used with a combination of transect external key and section external key.

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit,

I have had a quick look at the data to get an idea of some initial questions I can ask you.

1. Firstly, is it OK for me to attach your data file to this Github thread?
Note that Github threads are viewable by anyone, even if they are not a member of Github.

2. Are any of the data rows already likely to be on the EBMS system? or are they definitely all new rows?

3. Am I correct in thinking the people mentioned in the participant column are not users in the EBMS database already? or is it possible some of them do use the EBMS website themselves?


I can see the main difficulty with the import is something that Jim already mentioned to me in an email.
The section ID and name is not unique.
This means to know what section to attach a record to, we must check both the section ID and the transect ID together as a combination.
Unfortunately importer I will use does not support this way of doing things.
I will have a think of how to best to go about resolving this problem. 
It is slightly tricky at the moment as I don't have direct access to the database.
You do not need to worry about this last point I have made for now, it is just a note I am making for myself, I will try to find a solution.

**Comment by fontainebenoit** [comment]
Hi Andrew,
You can attach the datafile to the Github thread.
I don't think that any of the data rows are already on the EBMS system : they were downloaded from the former database, and I would be surprised if any of the participants had uploaded their data on both system.
However, it is probable that some of the people mentioned in the participant column are in the EBMS database already : these would be people who have been monitoring a site for several years, using the former database before 2023, and the EBMS since 2023. In this case, you could compare their transects in the file I gave you and in the EBMS database.

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit,

Thank you for your response.
This makes sense.

What I will attempt to do is attach the records to the appropriate users as best I can.
It may be the case that some records do not get attached, this is because people might use different names, or different variations of their own name on the website in camparison to the the upload file, however I will try my best.
Corrections can be made afterwards if absolutely needed.

Also thank you for the information about records not on the system.
I will see if can find an easy way to do some double-checks for this.
Again it is a little tricky as the data in the file may be presented differently to what is already present on EBMS.
Having said that, I think the locations in your file were imported by Jim, so that does suggest the data is not already present.

I do have a plan now, so can begin to proceed.
There are several steps to take, but I will let you know how I get on.

Andy

**Comment by andrewvanbreda** [comment]
Data file in question is here for reference
[French_eBMS.csv.zip](https://github.com/BiologicalRecordsCentre/ABLE/files/14527618/French_eBMS.csv.zip)

**Comment by andrewvanbreda** [comment]
Hi @BirenRathod ,

Would you be able to run the SQL in the attached file against live warehouse whenever you get a chance please.

I think you will need the following steps.

1. Unzip the attached data file onto Warehouse 1 at a suitable location.

2. In my SQL, change the path in the COPY statement to the path of where you placed the file.

3. Run the SQL.

4. Export the contents of the tbl_unique_section_ids_and_usernames table and send to me via email (as will contain username information).

5. 
I guess run this at the end to drop the temp table
drop table tbl_unique_section_ids_and_usernames;

If you have any questions, probably best to email me directly so we don't fill up the thread too much.

If the file can't be placed on Warehouse 1 directly, let me know, and I will dig out the syntax for the alternative Postgres command which I think allows copy from a local machine.

Thanks

[Temporary table SQL.txt](https://github.com/BiologicalRecordsCentre/ABLE/files/14558913/Temporary.table.SQL.txt)

[French_EBMS_reduced_columns_2.csv.zip](https://github.com/BiologicalRecordsCentre/ABLE/files/14558916/French_EBMS_reduced_columns_2.csv.zip)

**Comment by BirenRathod** [comment]
@andrewvanbreda See my email. The CSV has now been imported. Sorry I didn't read the 4 point. I have now sent you a file through email.

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit,

Now I have received the data I needed from Biren, I have been looking at preparing the data again.

However this has resulted in quite a few questions.

Questions are numbered for clarity.

1. 
I do not think the "id_obs" and "id_session" columns relate to anything on the EBMS system.
Am I correct in thinking these columns should not be imported?

2.
What is the rang column?
Again, does this even need importing?

3. The EBMS system uses Start Time, End Time fields.
Am I correct in thinking that in your spreadsheet, the "hour" column is equivalent to Start Time, and duration should be added to this to make the End Time
So for the first row we would have
Start Time = 17:07
End Time = 17:17
(duration on that rows says 00:10:00)

4. What is the Abundance Max column?
Does this even need importing?

5. I am not quite sure how to import the Clouds column.
On EBMS we just use a single percentage number, but I notice you data is a range of 2 numbers, for example
"nuages 51-75%"
I guess there are 3 options in this case
A. Use the lower number 51%
B. Use a number in the middle 63%
C. Use the upper number 75%

6. The same issue with the temperature field. Should we import using lower number, upper number, or in the middle. So for example, the first row is 26-30°C. We could import this as
A. Use the lower number 26
B. Use a number in the middle 28
C. Use the upper number 30

7. I am not sure how to convert the Wind column into something that can be used with the two Wind Direction/Wind Speed fields on EBMS. They fields seem very different to what is in the import data.
If you are unsure, then perhaps I can ask David for his opinion on this one.

8. What are you wanting done with all the Habitat columns?
I notice again they don't really match what we store on EBMS.
However they could be added as free text to the notes field I think?
What do you think?
If you think they should be imported as free text, do you have any opinion on whether they should be imported onto the Walk notes (sample)?, or the transection/section notes (location)?.
Again this is perhaps one I can ask David about

9. Am not quite sure why the lat/lon column does not seem to quite match the lat/lon I am seeing on the sections in the database.
This might be something Jim might be able to answer.
Perhaps it is to do with some lat/lons being in degrees/minutes, and others being decimal.
For instance in the first row of the spreadsheet we have the following lat long
-1.751591,43.377848
The section in the current EBMS database has these values instead
-1.75138 43.37757

I think that is enough questions for now!

Let know what you think, even if it is that you don't know the answer in some cases.

Thank you

Andrew

**Comment by fontainebenoit** [comment]
Hi @andrewvanbreda 
My replies below:

1. id_ob and id_session are unique keys for observations (ie species/abundance/date/time/place) and sessions (ie date/place) respectively. I suppose the eBMS system already has keys for these, but maybe it would be wise to import id_obs, which would allow to compare data in the eBMS and in the old database, just in case.
2. rang deals with the taxonomical rank of the observed taxon : species, subspecies, genus... The taxon name is explicit, so maybe you don't need to import it.
3. You are right
4. No need to import it
5. Use the number in the middle
6. Same
7. There is no wind direction in the former database, only wind speed in Beaufort scale, together with a speed range in km/h.
8. Yes, ask David what he thinks, but I would add the habitat description in the walk note, since they may change from one year to another.
9. I am not sure about this one...

I hope all this helps. Let me know if you need more info.
Best
Benoît

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit
Sorry I missed the notification for your message because of a problem with my email box.
I will take a look at the import again this coming week and let you know.

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit,

Apologies for the slow response.

Thank you for the information which will help me get closer to creating a file that is formatted in way that is suitable for import.

I think I am fine with the answers to the following questions, so these don't need any further comment unless working further on the file format creates further questions.

2 (rank would be useful, but only if the species isn't present at all in the database yet and we need to create it)
3
4
5
6
9 (I will ask Jim if I don't find an answer)

@DavidRoy, would you be able to comment on the following three points please?

Point 1 from the discussion list.
I think the "id_ob" column is equivalent to our occurrence ID. The Indicia importer will obviously generate its own. However, the existing one from the import file needs to be stored for future reference. One option is in the occurrence comment database field, as creating an attribute for something so specific to the France data is probably overkill. Although if you think other data that might be importer in the future will have this as well, it might be worth creating an attribute.

In terms of point 7.
I think the direction can be left as it isn't a mandatory attribute.
David do you know how the Wind Speed options on EBMS match up in regards to the Beaufort scale?
I notice the first 6 categories defined by the Met Office look similar to the 6 categories EBMS has
https://www.metoffice.gov.uk/weather/guides/coast-and-sea/beaufort-scale
So these could be matched up (e.g. Beaufort 1 could match EBMS Wind Speed category 1), or do you want to use a different rule to convert?

For point 8....David could you confirm your thoughts on what to do with the habitat columns?
One option is to import into the notes field on the sample as free text concatenated together?

Thanks

Andy

**Comment by DavidRoy** [comment]
@andrewvanbreda in response to your questions
1. including id_obs in the comment field is a good option
2. the wind speed categories are adapted from Beaufort scale so your suggestion is right
3. I agree on your suggested appraoach to the habitat information. We could extend the system in future to capture changes in habitat categorisation but currently we only have a 'static' habitat coding for transects

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit,

Apologies for the slowness of this issue, however it hasn't been forgotten.

I have formatted the data as best I can so far.

The most complex bit will be to work out how the rows split into samples and sub-samples like EBMS has, but I will worry about that once the data in the flat file has been formatted completely.

I have attached a small example of what the first 100 rows currently look like so you can see.
Note that not all the columns in the file will be imported. An example is the habitat columns which you will see are now in the sample comment, and the Duration which has been used to generate the End Time.

I contacted Jim about the lat/long values, but he didn't know either.
So what I have done is put them into the comment field (which is encoded using JSON).

I do have a few questions so I can finish the data formatting.

1. In the abundance column I sometimes see the word présence, however on EBMS our abundances are numbers. So I am not sure what to import these as? One possibility is 1. Any thoughts?

2. In the temperature column I am sometimes seeing ">35"
This is OK apart from on EBMS we use temperatures up to >40
I do not know how to import these in the best way.
Again, any thoughts are appreciated.

3. Final question for now.
I have converted most of the cloud values to a percentage number by using 0 for clear sky, and the middle number for most of the other values.
However I am not sure what to do with "Soleil voilé".
From what I can work out this translates to "Obscured sun". Do you have any thoughts of what this could be as a percentage?

Thank you

[EBMS France - Small Example.csv](https://github.com/user-attachments/files/15926655/EBMS.France.-.Small.Example.csv)

**Comment by fontainebenoit** [comment]
Hi Andrew

1. Can you put 1+ instead of 1 ? Presence means that the species is there, but not counted, so 1 is misleading. Or if you cannot put 1+, something like 9999 ?
2. Can't you just leave >35 ?
3.  Soleil voilé means that there is a light layer of clouds, but you still see the sun. So it's 100%, but very thin. Maybe translate it as 50% ? Anyway, I am not sure this is very important, since we tend to use offical meteorological data when needed.

Does all this make sense?
Best
Benoît

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit 
1. Unfortunately I cannot do this as it is a whole number only field, although I do recognise your point about it being misleading.
It is a limitation of the currently system that only exact counts seem to be allowed.
I will have a think of a solution.
2. I cannot put >35 because it is a set list of terms/words.
I could add a new term/word, but then other people would see it unless it is hidden from website and app.
Again, I will have a think.
3. Yes 50% sounds like a good compromise.
I can add a note to the sample comment also to say it is thin layer.

Thanks for your responses

@DavidRoy If you have any thoughts let me know

**Comment by andrewvanbreda** [comment]
Hi,

Further to my note above, I have done some more work on the format of the import files.

To briefly recap, I am aiming to create the following data structure on the EBMS website from the import file,
1. The main sample (this is the transect walk)
2. The subsamples (these are the section of the walk, attached as children under the transect walk)
3. The species occurrences which are attached to the subsamples.

The subsamples and occurrences are easy to organise, as the import file is one species per row, and the sections are clearly shown by the id_section number.

So it is the main walk sample I need to sort out, which I assume would be one walk per id_transect,recorder,date combination.

However there is a problem I can see with this.
In EBMS, we save the following relevant attributes once against the transect walk. However in the import file, these appear to be recorded individually against each section.

Start Time
End Time
Temperature
% cloud
Wind Speed

I do not think this is a problem for the Start Time/End Time, as we can store the earliest Start Time and latest End Time against the walk, then just store the exact times against the section sample comments.

However, the temperature, % cloud, wind speed also vary accross the transect on a specific date.

A specific example:
id_session 107013
This session is on date "26/08/2021", id_transect "11529".
Several cloud values are used and several Wind values.

Am not quite sure how to handle this.

The best thing I can think of is to take an average for the main sample walk attributes, and then put the precise values in the section subsample comments.

The only other alternative I can think of is to save each section with its own parent walk, but I think this is misleading and not the intended use of the EBMS data structure.

Any thoughts welcome.

**Comment by andrewvanbreda** [comment]
Also note:
I spoke to Gary van Breda on the phone the other day about the 2 unresolved issues noted above my comments from today.
He agreed that both the temperature and abundance issues did represent a problem for the EBMS system, as there is no equivalent in the current EBMS data model.

My own thoughts on these two issues.

1.
In regards to the temperature, perhaps we could simply not include the temperature attribute where the temperature is set as >35.
Instead, add a note of the value to the sample comment.
I notice this field is not mandatory anyway.

2.
For the occurrence, where the abundance is marked as present/présence (where there isn't an abundance number).
Perhaps I could suggest importing the occurrence without the adundance value, and adding a note to the occurrence comment.
This might mean these records might not be suitable for editing on the EBMS website, but they should be viewable I would have thought (although if any SQL reports join to the abundance attributes they might need changing to Left Join).
We could do a small experimental import.

These suggests might not seem ideal, but I am sturggling to think of ideal solutions.

If the above two solutions are not acceptable, perhaps another possibility would to simply exclude rows that are going to cause problems for the import.

**Comment by DavidRoy** [comment]
@andrewvanbreda thanks for looking into this.  I suggest we don't try to shoe-horn in the non-standard data into the same survey. The BMS method requires count data for example. I suggest remove the data that doesn't fit from this import. 

Then, as resources allow, the non-standard data could be imported through a separate survey? but this would need additional funding.

The solution for the temperature data is ok, as it's main purpose is to check that the weather conditions were suitable for the method and (to-date) has not generally been used for data analysis

**Comment by andrewvanbreda** [comment]
Hi @DavidRoy @fontainebenoit ,

Although there have not been any Github posts on this lately, I have working on this part-time and made quite a lot of progress with this now. I have organised the data to be what I think might be quite close to what can be imported.
I have attached the files I have been working on so you can optionally look at them if you so wish to do so.

Essentially though, there is a SQL file that reformats the data so that it is compatible with EBMS, and then it splits it into 3 separate import files which are intended to be imported one after the other.

- Samples (transect walks)
- Subsamples (the sections, which attach to the samples)
- Occurrences (species observed, attached to the subsamples/sections)

The system assumes that a transect walk is a unique combination of id_session, transect, date, this is what is used to group the samples

The SQL file contains comments on what the code is doing in case you want to know.

I do still have some questions though, which are below:
(I think these questions are probably mostly aimed at David to answer).

1.
I notice there are roughly 4000 rows where the abundance is 0.
Shall I just exclude these from the import?
We do have the concept of importing a zero abundance occurrence, however I notice on the EBMS website if you enter a zero, actually no occurrence is created at all, not even a zero abundance one.

2.
Again, in reference to the abundance, I notice some of the values are very large (example Abundance 10000 on the row "id_obs" 135277).
My understanding of EBMS is the abundance field is an exact count. I assume an abundance of 10000 is not an exact abundance count.
Is it OK to enter these values into the EBMS abundance field exactly as they are shown in the import file?

3. 
As previously discussed, some rows will not be imported due to incompatibilities with EBMS' expected data structure (e.g. where the abundance is just listed as "présence" instead of a number).
In these cases I have removed the rows from the entire session (indicated by the id_session number in the import file).
My intention is to avoid importing partial data, let me know if you prefer me to just exclude individual rows where the data is invalid (so data would be import partially from a lot of sessions)
When I exclude entire sessions with some invalid data, then about 20,000 rows are excluded (so about a 10th of the data)

4.
This question is probably for David only.
What verification/record_status do you want this data imported as?

5.
I have put an extra question in the attached file "Question 5", as it is a bit longer and involved, this is one for David.

[Question 5.txt](https://github.com/user-attachments/files/16635870/Question.5.txt)

[Data Formatting SQL.txt](https://github.com/user-attachments/files/16635880/Data.Formatting.SQL.txt)

[Import files (3 files).zip](https://github.com/user-attachments/files/16635900/Import.files.3.files.zip)

**Comment by fontainebenoit** [comment]
Hi Andrew 
My replies below. 
Thank you for this hard work 
Best 
Benoît 

> De: "andrewvanbreda" ***@***.***>
> À: "BiologicalRecordsCentre/ABLE" ***@***.***>
> Cc: "Benoît Fontaine" ***@***.***>, "Mention" ***@***.***>
> Envoyé: Vendredi 16 Août 2024 12:08:52
> Objet: Re: [BiologicalRecordsCentre/ABLE] French BMS - import samples and
> occurrences (Issue #658)

> Hi,

> Although there have not been any Github posts on this lately, I have working on
> this part-time and made quite a lot of progress with this now. I have organised
> the data to be what I think might be quite close to what can be imported.
> I have attached the files I have been working on so you can optionally look at
> them if you so wish to do so.

> Essentially though, there is a SQL file that reformats the data so that it is
> compatible with EBMS, and then it splits it into 3 separate import files which
> are intended to be imported one after the other.

>     * Samples (transect walks)
>     * Subsamples (the sections, which attach to the samples)
>     * Occurrences (species observed, attached to the subsamples/sections)

> The system assumes that a transect walk is a unique combination of id_session,
> transect, date, this is what is used to group the samples

> The SQL file contains comments on what the code is doing in case you want to
> know.

> I do still have some questions though, which are below:
> (I think these questions are probably mostly aimed at David to answer).

> I notice there are roughly 4000 rows where the abundance is 0.
> Shall I just exclude these from the import?
> We do have the concept of importing a zero abundance occurrence, however I
> notice on the EBMS website if you enter a zero, actually no occurrence is
> created at all, not even a zero abundance one.
OK for me to exclude these records 

> Again, in reference to the abundance, I notice some of the values are very large
> (example Abundance 10000 on the row "id_obs" 135277).
> My understanding of EBMS is the abundance field is an exact count. I assume an
> abundance of 10000 is not an exact abundance count.
> Is it OK to enter these values into the EBMS abundance field exactly as they are
> shown in the import file?
Yes 

> As previously discussed, some rows will not be imported due to incompatibilities
> with EBMS' expected data structure (e.g. where the abundance is just listed as
> "présence" instead of a number).
> In these cases I have removed the rows from the entire session (indicated by the
> id_session number in the import file).
> My intention is to avoid importing partial data, let me know if you prefer me to
> just exclude individual rows where the data is invalid (so data would be import
> partially from a lot of sessions)
> When I exclude entire sessions with some invalid data, then about 20,000 rows
> are excluded (so about a 10th of the data)
You should not remove the entire session : when there is a "P", or "présence", it means that the species was seen outside of the transect walk (for instance before starting the transect, or outside the 5 meter "box"). So please remove only the "présence" records, but not the entire session. 

>     1.

> This question is probably for David only. What verification/record_status do you
> want this data imported as?

> I have put an extra question in the attached file "Question 5", as it is a bit
> longer and involved, this is one for David.
OK for your averaging solution.

**Comment by DavidRoy** [comment]
I agree that the averaging for the section cloud data is ok as a solution

**Comment by andrewvanbreda** [comment]
Hi @DavidRoy @fontainebenoit , 

Thank you for your answers.

- I have changed the SQL code so that sessions (id_session) are still imported, even if one or more of the rows is invalid. Just the individual rows are removed.

- I have changed the code so that rows with abundance 0 are not imported.

- With the above changes (using tests on my machine), I note that 200286 occurrence rows will now be imported, so 7322 are excluded (which is less than before).

- Averaging will be used on Clouds, Wind, Temperature as mentioned.

I will do some more testing, and then do some small imports to start.

I think I just have one outstanding question, what status should these records be imported as?

I notice on a previous sample import for Spain I did for the Spring project (https://github.com/BiologicalRecordsCentre/SPRING/issues/117) have these fields set to
Release status = Accepted
Record status = Data entry complete/unverified

I assume the same here?

(note I am uploading the latest SQL file again for safe keeping, you don't need to check this again, unless you wish to look at it).

[Data Formatting SQL.txt](https://github.com/user-attachments/files/16662304/Data.Formatting.SQL.txt)

**Comment by fontainebenoit** [comment]
Hi @andrewvanbreda 
Yes, I assume the status of the French records should be the same as those from Spain.
Do you have an idea on when the import will be completed ?
Thank you in advance
Best
Benoît

**Comment by andrewvanbreda** [comment]
Hi Benoît (@fontainebenoit),
Thank you for the status information.
I am in a position where I can import a few records as a test, if it then goes well then I can import the rest of the records.
The reason this hasn't been done is because is not actually  a technical one, but is simply due to trying to do tasks in a priority order as best I can.

If you have a need for this import to be done very quickly please let me know, and I will move it nearer the top of my task list.
At the moment my plan was to try and do it within the next few weeks.

**Comment by fontainebenoit** [comment]
Hi Andrew @andrewvanbreda 
I have requests for data extractions for various areas and timeframes, which will be much easier to fulfill once all the data are in the database.
So even if it's not the highest priority, the sooner the better for me.
Let me know what you can do.
Thank you again
Benoît

**Comment by andrewvanbreda** [comment]
Hi @ @fontainebenoit
OK, I understand.
I will attempt to do a test import of a few of the records by the end of the week, then we can check the result if it works.

**Comment by fontainebenoit** [comment]
Great, thank you !

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit ,

OK, so I have managed to do a small import of data (about 120 occurrence rows), however there are quite a few difficulties still to overcome that will definitely mean I cannot do the full import very quickly.

I have attached the import files that were used.

I think the data has imported ok, although I need to check in further detail, it is quite a complex task to really be sure.
I would be interested in your opinion of the data.
Obviously, the first thing to check is to make sure you can see the data? Let me know if you can see it.

I had to sort the data in a particular way for my import, so the attached files are not in the same order as your original data file.
However, if you open the occurrences files you can see the id_obs number in the occurrence comment column which will tell you which rows were imported.

There are several problem I experienced during the import, which meant I had to manually change the files to get them to work.
Obviously I cannot do that when we do the final import as there are too many rows, so I need to find solutions to continue with the import.

One of the problems is particularly difficult to solve, but I will try to find a way, I will not go into detail about this (unless you want me to) otherwise this Github thread will be too long.

The other main difficulty is about the species themselves.
I found 12 of the species entries were not found in the EBMS Butterflies list at all.
I assume this will require me to query the entire species column in the original data file and compare with EBMS Butterflies species list. I can see in some cases it is because my code has failed to remove the authority from the name, in others it may be the way the species is written "Thymelicus et apparentés" means the match isn't found.
After doing that, get any missing entries added to EBMS Butterflies species list if David is OK with that.
It was about 1 in 10 rows that failed in this way (file containing the failed rows is attached)

Let me know if you can see the data and if it seems roughly ok to you?
[1. Samples (small).csv](https://github.com/user-attachments/files/16994694/1.Samples.small.csv)
[2. Subsamples (small).csv](https://github.com/user-attachments/files/16994697/2.Subsamples.small.csv)
[3. Occurrences (small).csv](https://github.com/user-attachments/files/16994698/3.Occurrences.small.csv)
[1726214444406-errors (occurrence errors).csv](https://github.com/user-attachments/files/16994699/1726214444406-errors.occurrence.errors.csv)

**Comment by fontainebenoit** [comment]
Hi @andrewvanbreda 
I can see the data. They seem OK at first glance.
For the species problem, it seems to be often due to the fact that we have species groups in the French BMS
 : "Lycènes bleus" is "Blue Lycaenids", "Thymelicus et apparentés" is "Thymelicus and related species" etc.
I am not sure how you can deal with these. I have a file which gives, for each species group name, the species which are included in the group. Would that be helpful ?
Or you could send me the species entries which cause problems, together with the EBMS species list, and I will give you the correspondence.
If you don't have species groups in the EBMS, then we will have to leave them, no?
Let me know what you think.
Best
Benoît

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit,

Yes it would be good idea for me to query the database and then get the species that don't match for your opinion on them. I think a few of them I will be able to fix myself, but definitely not all.

Thank you for taking a look at the data to make sure it is ok at first glance. This was my opinion too when I checked it.
I think the structure of the data should be ok, but I will double-check the averaging calculations I previously mentioned worked ok.

Anyway, I will let you know how I get on.
Yes feel free to send me anything you think might be helpful.

Andy

**Comment by fontainebenoit** [comment]
[STERF_Liste_Rhopalo_France.pdf](https://github.com/user-attachments/files/17063708/STERF_Liste_Rhopalo_France.pdf)
Hi @andrewvanbreda 
Please find attached a file with the names of species groups and the species which are included in each of them. I think it is self-explicit, but if not, don't hesitate to ask for explanations.
Best
Benoît

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit ,

Thank you for the species information.

I can't comment too much on this at the moment as I have not been able to run the query to check the species in the original file against the database yet. 
To do this the file data needs to be copied to the server, but I have not got the required access to be able to do this at the moment, hopefully that will be sorted in the next few days and we can progress the import again.

Once I have done this the file you just sent will likely be very useful.

Andy

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit ,

I managed to find a way to query the species against the database.

I found about 100 problems out of the 300 distinct species from your original import file.

I fixed about 50 myself, as they were actually related to the taxon name and authority not being separated properly by my import code, so there wasn't really anything wrong with the actual species name.

The remaining 42 are in the attached spreadsheet.

I added some notes to the spreadsheet as I was going along, but I have removed them for now, as my ideas about how to solve the problems may change depending on what you say about the examples below.

I see that the document you sent the other day may have some explanations on how to fix problems if I can correctly understand it.

Once I understand it I will send you my full notes on how I propose to solve issues with species in the import file.

I will give 2 examples to help me understand it.


EXAMPLE A
Taxon "Pieris à nervures visibles" from the original import file.

Is my understanding correct that the document is saying this could be Pieris napi, or Pieris brygoniae, but the recorder did not specify which?

If so, we need a way to record that.

I can only think one of 3 options for this particular example

1. Create an actual Pieris Nervures Visibles taxon to record against.
2. Save the occurrence as Pieris but with "à nervures visibles" in the occurrence comment
3. Save the occurrence as Pieris, but with "could be Pieris napi or Pieris brygoniae" noted in the occurrence comment.


EXAMPLE B
In the import file there is a taxon "Clossiana et Boloria"

If I look in your document, there is reference to "Boloria, Clossiana, Brenthis ..."
This is coloured grey.
I see that next to this is Issoria Lathonia which is also coloured grey.
Should I therefore conclude that Issoria Lathonia is the species when "Clossiana et Boloria" is shown in the original import file?

Thanks for any thoughts

Andy


[Species names that are not present (notes removed).xlsx](https://github.com/user-attachments/files/17133375/Species.names.that.are.not.present.notes.removed.xlsx)

**Comment by fontainebenoit** [comment]
Hi @andrewvanbreda 
You are rigth for "Pieris à nervures visibles", it relates to Pieris napi or Pieris bryoniae. The occurrence should be saved as "could be Pieris napi or Pieris bryoniae ("Pieris à nervures visibles" pseudotaxon)".
However, you have been misleaded by the colors for the "Clossiana et Boloria" group. Colors work only for the third and fourth columns of the table. This group represents 13 species, ie between Issoria lathonia and Clossiana dia in the table.
Does it make sense?
This occurrence should be saved as "Small fritillaries", with a comment "Includes all species in genera Issoria, Brenthis, Boloria Proclossiana and Clossiana".
That means that you would have to create a new taxon. Would that be possible?
Let me know what you think
Best
B.

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit ,

OK, these makes sense.

However, as I am just a programmer, so I won't be able to decide on what entries would be deemed acceptable for the EBMS Butterflies species list. These data quality issues will need to be decided by David.

@DavidRoy would you be able to comment if you wish to allow these?

I have re-attached the list of affected species, but this time with comments. A lot of the comments say that it will be a pseudo taxon dependent on your STERF_Liste_Rhopalo_France document, but not all the comments are like that.
I have changed the order of the list so all the ones not like that are near the top of the list.
Benoît: you may or may not wish to comment further on what I have said.

I have also re-uploaded the STERF_Liste_Rhopalo_France.pdf document so we don't have no navigate up the Github thread to find relevent documents.


[Species names that are not present in database.xlsx](https://github.com/user-attachments/files/17135559/Species.names.that.are.not.present.in.database.xlsx)


[STERF_Liste_Rhopalo_France.pdf](https://github.com/user-attachments/files/17135501/STERF_Liste_Rhopalo_France.pdf)

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit,
I have mailed David as he might be swamped with Github notifications and does not see this.

One issue I thought of is that it might be a problem if other people can see these extra species during their own data entry, but I will let David decide that. 
One possibility is that is a "Allow Data Entry" flag that could be switched off to possibly avoid that, but I am unsure of the implications of doing that.
It is difficult for me to comment on the right approach as I am not familiar with the entire system (such as any phone apps).

I will try to sort out some of the other problems encountered during the test import while we wait, there is one in particular which is tricky.

**Comment by fontainebenoit** [comment]
Hi @andrewvanbreda 
Please find attached my comments.
Let's see what David thinks about creating new pseudotaxa. It would allow to carry on with the same taxon/pseudotaxon list as we used to do in France. However, I think there is a small proportion of entries in the database that relates to pseudotaxa (could you check?). So if we don't add them, it's probably no big deal.
Best
[Species.names.that.are.not.present.in.database.xlsx](https://github.com/user-attachments/files/17150807/Species.names.that.are.not.present.in.database.xlsx)

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit 

Actually you are right, there is not too many, I checked and it roughly about 3500 that would not be imported out of over 200,000.
Given this, and that I have no idea about David's availability at the moment, I will investigate whether it is easy for me to organise the import such that these rows can be imported later.
This will be better, otherwise we can get in a situation where we are trying to fix so many issues we never actually do the import.
I have re-attached the notes file, there are 4 new comments in bold. 3 of these just confirm the presence of the suggested synonym.
The one question relates to Quercusia quercus which I think might be Favonius quercus on the system.


[Species names that are not present (updated).xlsx](https://github.com/user-attachments/files/17151915/Species.names.that.are.not.present.updated.xlsx)

**Comment by fontainebenoit** [comment]
Hi @andrewvanbreda 
Yes, Quercusia quercus is a synonym of Favonius quercus.
The question of synonyms may be a problem for the participants: since the accepted name in the French taxonomical reference list (Taxref) is Quercusia quercus, french lepidopterists usually use this name. Will they be able to record observations of Quercusia quercus under this name ? I think it is important that both names are available for recorders. In the database afterwards, both names should point to the same taxon. I guess you already had that synonym issue, but I point it out just in case.
For the pseudotaxon, I agree with you. We can leave the problem for now if David is not available, and import only the other records, relating to real taxa.
Best
Benoît

**Comment by DavidRoy** [comment]
Thanks for your work on this @andrewvanbreda @fontainebenoit 
A couple of immediate comments but let me know if you need views on specifics
1. adding complexes is fine.  Could you enter them in the format of existing complexes for the eBMS butterfly list. e.g. https://github.com/BiologicalRecordsCentre/ABLE/issues/678
2. adding synonyms is also fine when these are clear-cut. We add synonyms on a case-by-case basis so our current list is not extensive
3. we align with the published european butterfly list https://zookeys.pensoft.net/article/28712/ so we might need to align some species concepts? If needbe, we could liaise with others in BCE to get agreement where there are different views on this
4. we don't typically include infra-specific taxa (subspecies/forms etc) but it's ok to add them if you have a strong reason for this

Is that enough to complete the work on this?

**Comment by DavidRoy** [comment]
Our posts crossed. @andrewvanbreda do the website input forms allow lists to include synonyms?
@kazlauskis do you include synonyms in the app species lists?

**Comment by fontainebenoit** [comment]
Thank you for your replies @DavidRoy 
Regarding subspecies, we don't have a strong reason to add them. There are very few occurrences of subspecies in our database, so it's ok for me to import them at the species level.
@andrewvanbreda, let me know if you need more input from me.

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit @DavidRoy,

I am working on going through the species, and I will provide a new organised list that includes proposed solutions (include the actual names of complexes I will add that you can confirm)

I am realising some of the species we previously thought might need adding are actually already present in the system under a synonym name, so there are going to be fewer pure species to add than originally thought.

This new list won't be finished today.

However just to answer your question about the synonym names and website input.

Although I have experience of the data entry form, I did not write the form originally, so am going purely by observation.

What appears to happen is that any name at all (including common and synonyms) can be used to search in the "Search for or add species to list (Butterflies)" box.
However when this is added to the grid, the system displays and saves the preferred name to the database.

I think my conclusion regarding this is as follows (ignoring the app)...
You should be able to use your synonyms for entry, but when they are shown on screen or in downloads, the preferred name will be shown.

**Comment by fontainebenoit** [comment]
Hi @andrewvanbreda 
The way you manage synonyms is fine, thank you !

**Comment by andrewvanbreda** [comment]
Hi @DavidRoy @fontainebenoit ,

Please find attached a spreadsheet containing my propose solutions for issues with the 42 species that have problems finding matches in the database.

For each row I have put some notes on the right.

Ultimately most of these falls into 2 categories.

Benoît: Would you be able check my interpretation of which species should be included in the aggregates please.
There are quite a few rows, so perhaps it is not necessary to check every row if you think I have got the right idea.

David: Quite a few rows I have had issues with the name of the aggregate, often this is due to its length.
Would you be able to let me know your thoughts on how to handle these please?
Again there are quite a few rows, so perhaps the best thing might be best to let me know your general thoughts, rather than put the exact solution on each row.  There is also one case where I wasn't sure as it was a Genus but without two specific species.
I did ask John, there are some ideas in the from him near the top.
There are also some issues which are unique to particular rows, I have tried to put these near the top also.

There are also some solutions that I don't think need further input from you. I have put these at the bottom, feel free to read through these and comment if you so wish.

Note that most of my research comes from a combination of examining the warehouse and looking things up in google, so feel free to point out any inaccuracies. Where possible I have also tried make make sure we are not duplicating species already in the database, there are cases where I have found a synonym that could be used.

Thanks

Andy

[Species list issues with solutions.xlsx](https://github.com/user-attachments/files/17276474/Species.list.issues.with.solutions.xlsx)

**Comment by fontainebenoit** [comment]
Hi @andrewvanbreda 
My replies in the attached file. let me know if I can help you in a way or another.
Best
Benoît
[Species.list.issues.with.solutions.xlsx](https://github.com/user-attachments/files/17400259/Species.list.issues.with.solutions.xlsx)

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit (and @DavidRoy),

Thanks for these responses Benoît, really helpful.
I don't think I have any questions for you regarding what you have said.
I will need some further answers from David, I will send him an email in case he misses Github notifications.

@DavidRoy : could you let me know about the following in the spreadsheet? (see link in Benoît's comment above) Thanks

1.  The question about adding Melitaeini as a Tribe

2. Have you any thoughts on the long names?
It sounds like long names are a problem, Benoit is suggesting the use of the world "group", although I notice on the Warehouse sometimes the word "others" is used e.g. "Pieris rapae/melete/others".
I guess "group" is actually better, as it is more flexible (e.g. can include multiple Genus).
Once we agree on the name format to use, how do we capture the information regarding which species a grouped name refers to?
Does there need to be in a document somewhere as Benoit is suggesting?
I would also suggest perhaps putting the individual species name into the description on the taxon in the Warehouse (although I think that is probably not viewable by a user of EBMS).

3. Could you give me your thoughts on the line "Pyrgus autres que malvae/malvoides".
This is unusual scenario where we have a Genus but does not include 2 species.
Am not sure how to name this, could you also confirm Benoit's thoughts that the taxon rank should be Genus?

4. I assume you are ok for me to add the synonyms to the EBMS Butterflies list myself as Benoit suggests.

Thanks

Andy

**Comment by DavidRoy** [comment]
sorry for being slow on this one.  Responses:
1. ok to add the Tribe
2. I think the use of 'group' is ok for situation where the names are very long. Aggregates/complexes of 3-4 taxa are managable but otherwise use 'group'. I'm happy for you and Benoit to decide on these
3. I think this can be covered by 'Pyrgus' as a genus level taxa. If specific species (malvae or malvoides) are recorded then presumably the full species would be used?
4. Yes, happy for you to implement

In general, the taxon list is defined by the national scheme.  Thanks for asking my view and all the thought into this, but ultimately it's Benoit's decision for France

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit @DavidRoy ,

Just to let you know I have started adding the new names.

I do not think I have many more questions. 
I will see how I get on trying to create sensible group names and then we can refine the namez afterwards if needed. That will probably be quicker than us discussing what each group name should be before I add them.

The process of adding the new names isn't complex, but takes a little time as I also need to change the code that generates the import files to use the new names as I go along.

I notice there are a series of presence code attributes on the taxa in the Warehouse, I will set these to "P" for "FR".....as I assume it would be correct to set these new names are present for France. Let me know otherwise if needed.

Once the names are added, I will try to correct the previous test import that was done, as that missed out on some taxa name that weren't available at the time.

Then after that I can start looking at the possibility of importing all the data (which will be imported in chunks rather than all at once).
There was a bit of a difficulty when I did the test import relating to attaching the occurrences to the samples, so I cannot say exactly when I will be able to get the import working.
I will try my best to get it done as soon as I can though.

Just quickly in reference to your point 3 David.
Yes OK, I can set those "Pyrgus autres que malvae/malvoides" ones to be attached to just "Pyrgus" without the need for a new entry.
There are indeed other references to "Pyrgus malvae" and "Pyrgus malvoides" specifically in the import data, these occurrences will just simply attach to those exact names which already exist in the EBMS Buterflies species list.

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit @DavidRoy ,

I have added the new names to the EBMS Butterflies list.

I have attached the usual spreadsheet, but will the details of the added entries in an extra column on the right.

I think it was mostly OK, but you may want me to alter some of the group names as they are currently only what I personally thought might work.

Here are the group names I created:

Pieris and Leptidea group
Erebia epiphron group
Lycaenidae blue group
Pyrgus related group
Euphydryas/Melitaea/others group
Boloria and Clossiana group
Mellicta and Melitaea group
Plebejus and Vacciniina group
Lycaenidae red group
Thymelicus related group
Argynnis related group
Euchloe and female cardamines group

There are obviously all the other entries, but they are either the addition of a synonym or use standard aggregate naming format (details are in the attached spreadsheet if you want them)

I think I only have 2 questions:

1. 
I realised I made a mistake with "Petit".
What happened was I had been looking at the plain text of the original css data and I saw this “Petit," but I realise now the comma is not the separator to the next column, but part of the taxon name. The full name is this “Petit, à nervures non verdâtres (unies)”.
However I still don’t know what to do, as I can’t see it in the STERF_Liste_Rhopalo_France.pdf document.
Could you let me know which species should be include in this group please?

2. 
I wasn't quite sure how to name the "Euchloe + femelle de cardamines", in the end I just translated directly to “Euchloe and female cardamines group". However, I am used the species sex being an attribute rather than in the taxon name, let me know if this needs renaming.

(Benoît, you will see in the spreadsheets links to the new entries in the Warehouse, I guess you don't have a login to the Warehouse, but hopefully there is enough information to see what is going on without needing to use the links).

Thank you for any assistance



[Species list issues with solutions.xlsx](https://github.com/user-attachments/files/17521977/Species.list.issues.with.solutions.xlsx)

**Comment by CrisSevilleja** [comment]
Hi @andrewvanbreda just a question. Are all these groups included in the eBMS species name list or just applied when selecting France? Thanks for all the work for matching French records

**Comment by andrewvanbreda** [comment]
Hi @CrisSevilleja 
I did just check the EBMS website walk data entry form, and they are available for selection there.
When creating the groups I did only select the Presence attribute for France only, however I do not know what those attributes actually do because I did not program the website originally, so my knowedge is limited in some areas.

I don't think they would be available in any apps at the moment though, as I think species data needs importing into the app from the Warehouse, the app doesn't get its species lists in realtime.

**Comment by CrisSevilleja** [comment]
thanks for the information @andrewvanbreda

**Comment by andrewvanbreda** [comment]
Progress notes on this issue

- The first test import I did (that was only partially complete because of missing group/species names) has now been completed.

- I have changed my SQL code to output import files that will use the new species names/groups

- I have fixed my SQL code to correct a few minor problems where manual intervention was required during the test import. Also changed code to exclude the data already imported.

- I have fixed 2 major difficulties that would have prevented the import from working.
One of these is a fix to the code the database server runs.
This code has been changed and has been committed. The new code is now just awaiting release onto the live server.
Once that is on there, I should be able to go ahead with the import I think.

Attached is my latest SQL code for safe keeping.

[Data formatting SQL (with initial test import data removed).txt](https://github.com/user-attachments/files/17623309/Data.formatting.SQL.with.initial.test.import.data.removed.txt)

**Comment by fontainebenoit** [comment]
Hi @andrewvanbreda 
I have some requests about French data. Do you have any idea about when the import will be complete? I just need to be able to give an answer to those people.
Thank you in advance
Best
Benoît

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit,
I am waiting for a code change to be put on the system before I can do the import.
However, the other day I thought of another possible way to do the import which may mean I don't need to wait for that, but I need to check it will work.
I will do that check at the weekend and let you know early next week if i am able to get it done quickly.

**Comment by fontainebenoit** [comment]
Great, thank you for your prompt reply!

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit,

I did the experiment I mentioned at the weekend and it did not work.
I also have a problem where I can't access the database directly at the moment, which means I cannot even use the slower technique I was able to use when importing the recent test records.

However, I do have one other option, I can fast-track that change I mentioned to the importer, and get it put on the system quicker than it would otherwise be.
The downside of this is it means that change will no longer be checked by anyone else, so I need to do some further testing to make sure it really is working perfectly.

Given this, are you able to tell anyone asking about data that it should be on the system in the first half of December. Let me know if that is too late.

**Comment by fontainebenoit** [comment]
Thank you @andrewvanbreda, first half of December would pe perfect!
Best

**Comment by fontainebenoit** [comment]
Hi @andrewvanbreda 
Any news about the import?
Thank you in advance for the update.
Best
B.

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit ,
I have been testing it on my machine and it appears to be working as best as I can tell in that limited environment (I don't have the species list, users on my machine).
So my intention is to start importing the data this weekend.
It should still be done in the first half of December timeframe.

**Comment by fontainebenoit** [comment]
Great, thank you!

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit 
Just a quick update on progress.
I manage to import about 5000 rows with no trouble with the structure of the data.
This is really good that is importing correctly without failing rows.
The speed of import also seemed to be ok, the full 200,000 would a few hours to import, so that isn't an issue.

The only problem is the verification status on the records, is not what I expected, and I am looking into this with another developer.
I will wait until tomorrow to see what happens with the verification status investigation, but if there isn't quick progress on that, I can import the records and alter the status after if that is the only way.
I still expect completion this week.

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit @DavidRoy,

I finished the import this evening.

I am not sure, there may (or may not) be a delay before the data appears on your download/report Benoît.
Have a look, but don't be alarmed if all the data is not present yet. If it isn't all there, I would be interested to know how much of it you are seeing.

The import seemed to go relatively smoothly considering it is 200,000 rows, so I hope the data is ok once you check it.

I encountered 3 different issues.

1.
As mentioned, the first 5000 of the occurrences didn't have the verification status I expected.
I will look to get this corrected.
This occurred because I misunderstood a setting, it is not a error in the importer.

2.
350 rows failed because they were linked to the taxon "Petit, à nervures non verdâtres (unies)", and this is not in the species list.
I note this issue was raised above, but I didn't notice it hadn't been resolved.
I don't think this is in the STERF_Liste_Rhopalo_France sheet.
If we can decide which species this relates to, it would still be easy to import these if I create a group name like before.

3.
I am not quite sure why this wasn't noticed before, but 2 records are linked to "Coenonymphina" and that is not in the species list.
From what I can work out by doing an Internet search. this is a Subtribe. However if I do a search in the species list for an example Genus such as Coenonympha in that Subtribe, I can see the EBMS Butterflies taxon hierarchy doesn't include the subtribe, so I can't just add it without the hierarchy becoming disjointed regarding items that would be below Coenonymphina.
So I am not sure what to do with this.
We could just import the rows as the next available taxon higher in the taxon hierarchy which I think would be Satyrinae.

**Comment by andrewvanbreda** [comment]
Note that there is no need to respond to the following comment. These are just some files used during the import so they can be referred to in case of any issues.
The 4 files are
- SQL file that organises the original data and creates 3 import files (samples, subsamples, occurrences)
- File describing the import settings used
- An altered version of the original data file more suitable for import (such as it has extra columns). (file is zipped for upload to Github)
- An extra data file also used by the SQL.


[Data formatting SQL.txt](https://github.com/user-attachments/files/18110483/Data.formatting.SQL.txt)

[France_eBMS_original_data_with_extra_columns.csv.zip](https://github.com/user-attachments/files/18110484/France_eBMS_original_data_with_extra_columns.csv.zip)

[Settings for data import.txt](https://github.com/user-attachments/files/18110489/Settings.for.data.import.txt)

[ebms_france_transect_section_srefs.csv](https://github.com/user-attachments/files/18110491/ebms_france_transect_section_srefs.csv)

**Comment by DavidRoy** [comment]
Thanks @andrewvanbreda. Ping me an email if anything specific but I'm happy to leave this to you and Benoit.
great job!

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit @DavidRoy 
I have corrected the verification status problem on the 5000 occurrences I mentioned. They are all now set as released and unverified.

Only point 2 and 3 above need resolving ("Petit, à nervures non verdâtres (unies)" and "Coenonymphina" issues)

David:
Note I realised the same problem had occurred on 2 imports done last year, so I have corrected these back to released, unverified too.
They were previously incorrectly set as verified by me as I misunderstood a setting. Not sure how much that matters now if that project is ending, but as I was sorting out the EBSM France data I corrected it anyway
SPRING/issues/117 (Spain import)
SPRING/issues/124 (Netherlands import)

**Comment by fontainebenoit** [comment]
Hi @andrewvanbreda 
I am working on the eBMS data. I have trouble linking the tables I download from the eBMS website. Can you tell me which are the keys between site details, transect sections, occurrences and surveys ?
Thank you in advance
Benoît

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit,

Ok no problem, I will try to explain my understanding of how the data is organised (I have number each point hopefully to make it easier to read my comments)

I am assuming you are asking about the download files themselves rather than how the data looks in the database tables.

I am assuming you are downloading your data from this page. Let me know if this is not the case.

https://butterfly-monitoring.net/downloads

I have logged in as your user, and downloaded the visits and occurrences data from this page so I can hopefully see what you are seeing.


Apologies, as some of what I say may already cover what you know, but it is easier to give a general overview.


1. 
(how to identify data that has been imported?)


Firstly in the visits file.
If you look at the "Comment" column, any row which includes a session number in the following format is part of the imported data
e.g.
"{"id_session": "104487"}"


2.
(how does the visits and occurrences file link together?)


Each row in this visits file is an individual transect walk, there is only one row per walk.

However in the occurrences download file there can going obviously to be multiple rows per walk.

The column that links the two files is "Transect Sample ID" number, this is present in both files. This is the unique number in the database for the visit.
There is only one row for each "Transect Sample ID"  in the visits file.
for example:
28173487 appears once in the visits file

However if we search the occurrences file, there are 34 rows related to this visit number, so there are 34 occurrences for this visit.



3.
(how sections are organised)


In regards to the section sub-samples, it is the "Section Sample ID" column in the occurrences file which is the unique number in the database which represents each section of the walk. 

(Note: This number is unique to the visit's section itself, so if another walk used the same section location, it would have a different Section Sample ID number).

As there can be multiple occurrences for a section, this means that each Section Sample ID can appear on more than one row in the occurrences file.
For example, we can see that Section Sample ID 28173821 appears 7 times, so 7 occurrences were observed on that section for that walk.

I notice in the download file, the actual location ID of the section is missing (the section equivilent of the unique "Transect ID" number column for the location).
As the file does not include a Section ID number for the location, if you want to identify a section location, you need to look at a combination of the Transect Name, Section Name (or Section Code) columns.

Note that when I say "Transect ID", it is a number generated by the database, it is not the same number as the "id_transect" in your original data file (although it is unqiue to the transect location in the same way as id_transect is).
Your own id_transect and id_section numbers are not included in the download file at the present time.
I am not sure if these were imported, as Jim Bacon handled the location import. I can look in the database if you need them.


4. 
(about averaged data)


If you remember when we discussed importing the data, I had to do an averaging of some of the data to get it to import in EBMS format.

For example, if you look at the visits file, the "Wind Speed" column is actually an average of your data. This is because EBMS saves this value against the whole visit, whereas your data has this data recorded per section.
The raw data from your observations can still be seen at the section level however, by looking in the "Comments" column in the occurrences file. This data is not averaged, it looks something like this.

{"hour": "10:30", "wind": "3 Beaufort (12-19km/h)", "clouds": "Ciel dégagé", "duration": "00:10:00", "latitude": "43.368371", "longitude": "-1.725962", "habitat 1-a": "Prairie cultivée", "habitat 1-b": "Entouré de haies avec arbres", "habitat 1-c": "Fauché", "habitat 2-a": "-", "habitat 2-b": "-", "habitat 2-c": "-", "temperature": "16-20°C", "code habitat 1": "4.1.1.2", "code habitat 2": null}

(This reason why the data looks like this is it has been formatted in a standard format called JSON).

The "Occurrence comment" column next to this contains your own id_obs number and is formatted in a similar same way
e.g.
{"id_obs": "183019"}


I think perhaps you will likely have some more questions after this answer. I hope my explanation is not too confusing, particular point 3 about the sections.

If there is any data you require that is not displayed in these downloads files, then let me know the data fields you are looking for and I will try to advise.

I hope that helps

Andy

**Comment by fontainebenoit** [comment]
Hi @andrewvanbreda 
Thank you for this detailed reply. I think I have most of the answers I need, but I will work on the data and maybe will have more questions later.
Meanwhile, I'd be interested to have my original id_transect and id_section numbers if they were imported, that might be useful (though not essential).
Best
Benoît

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit,

I can see that the id_transect and id_section were indeed imported into the transect/section locations in the database by Jim using the external_key database field.

However, I tried experimenting with my development site, and I couldn't get it to include the id_transect and id_section numbers in the download itself.
This is because the downloads rely on something called ElasticSearch (which is very fast) and the external_keys are not included in the ElasticSearch index at the moment.

The only thing I have been able to do is download the transections and section location records directly from the database, perhaps this is not to useful because it is just the locations, but I have sent them anyway.

In these files you will see a column called "transect_database_location_id" which matches the Transect ID field in your sample/occurrences download.

Another quick note as a reminder about the data.

If you remember when we examined your original import data file, we found that the latitude/longitude recorded on rows often differed slightly to what was stored in the location records that were imported into the database.
I think we never quite worked out why this was.

For example, I can see that sect "1032_1" has centroid spatial reference 47.51142 0.27406 in the database location record (and in the file i am sending you now) .
However the latitude/longitude in one of the rows in your import file for this section was recorded as this:
"latitude": "47.511335", "longitude": "0.27397"
If you are unsure, the actual recorded latitude and longitude can always be found in the JSON data in the Comments column in the occurrences download file as seen here,

{"hour": "16:30", "wind": "5 Beaufort (29-38km/h)", "clouds": "Ciel dégagé", "duration": "00:10:00", "latitude": "47.511335", "longitude": "0.27397", "habitat 1-a": "Prairie non cultivée", "habitat 1-b": "Entouré de haies avec arbres", "habitat 1-c": "Fauché", "habitat 2-a": "Forêt mixte dom. conifères (> 10% feuillus)", "habitat 2-b": "Clairière et lisière", "habitat 2-c": "Peu exploité", "temperature": "16-20°C", "code habitat 1": "4.2.1.2", "code habitat 2": "1.3.3.2"}

[France Sections.csv](https://github.com/user-attachments/files/18506497/France.Sections.csv)
[France Transects.csv](https://github.com/user-attachments/files/18506496/France.Transects.csv)

**Comment by fontainebenoit** [comment]
Hi @andrewvanbreda 
I have a question : we have BMS data which have never been uploaded in the previous database, representing several years of monitoring over several sites. I'd like to upload them in the eBMS. They are on excel files. Is there a way to import them all in one go? If yes, how should they be formatted?
Thank you for your help
Best
Benoît

**Comment by andrewvanbreda** [comment]
Hi @fontainebenoit ,

Is it possible to send me some lines from the files so I can see how the columns are formatted, then I can advise.
Also could you let me know how many rows are in the files in total.

It is the format of the data that is the important thing, the file type does not matter, we can easily convert Excel to what we would need.

**Comment by fontainebenoit** [comment]
Hi @andrewvanbreda 
Thank you for your reply. I have asked the guy you have the data to send me an example  and to tell me how many rows they have.
I'll keep you informed
B.

**Comment by DavidRoy** [comment]
@andrewvanbreda @fontainebenoit thanks for all your work on this. I think the original requirement is now finished so I will close this issue. If we want to import the other legacy data, I suggest a new issue is raised
